lemminx-liberty/src/main/java/io/openliberty/tools/langserver/lemminx/codeactions/AddFeature.java

```diff
@@ -0,0 +1,108 @@
+     *      1) Add a feature to an existing empty featureManager
+     *      2) Add a feature to an existing featureManager with children
+     *      3) Add a feature and new featureManager
+     *  
+     *  To calculate where to insert, each scenario will use a reference point to calculate range
+     *      1) The startTag of the featureManager
+     *      2) The last child of the featureManager
+     *      3) The startTag of the server.xml
+     */
+    public static final String FEATURE_FORMAT = "<feature>%s</feature>";
+    public static final String FEATUREMANAGER_FORMAT = 
+            "\n\t<featureManager>"+ 
+            "\n\t\t<feature>%s</feature>"+
+            "\n\t</featureManager>";
+
+    @Override
+    public void doCodeAction(ICodeActionRequest request, List<CodeAction> codeActions, CancelChecker cancelChecker) {
+        Diagnostic diagnostic = request.getDiagnostic();
+        DOMDocument document = request.getDocument();
+        TextDocument textDocument = document.getTextDocument();
+        String insertText = "";
+        int referenceRangeStart = 0;
+        int referenceRangeEnd = 0;
+
+        for (DOMNode node : document.getDocumentElement().getChildren()) {
+            if (LibertyConstants.FEATURE_MANAGER_ELEMENT.equals(node.getNodeName())) {
+                DOMNode lastChild = node.getLastChild();
+                if (lastChild == null || !lastChild.hasChildNodes()) {
+                    // Situation 1
+                    insertText = "\n\t" + FEATURE_FORMAT;
+                    DOMElement featureManager = (DOMElement) node;
+                    referenceRangeStart = featureManager.getStartTagOpenOffset();
+                    referenceRangeEnd = featureManager.getStartTagCloseOffset()+1;
+                } else {
+                    // Situation 2
+                    insertText = "\n" + FEATURE_FORMAT;
+                    referenceRangeStart = lastChild.getStart();
+                    referenceRangeEnd = lastChild.getEnd();
+                }
+                break;
+            }
+        }
+        // Situation 3
+        if (insertText.isEmpty()) {
+            insertText = FEATUREMANAGER_FORMAT;
+            DOMElement server = document.getDocumentElement();
+            referenceRangeStart = server.getStart();
+            referenceRangeEnd = server.getStartTagCloseOffset()+1;
+        }
+        Range referenceRange = XMLPositionUtility.createRange(referenceRangeStart, referenceRangeEnd, document);
+
+        try {
+            String indent = request.getXMLGenerator().getWhitespacesIndent();
+            insertText = IndentUtil.formatText(insertText, indent, referenceRange.getStart().getCharacter());
+        } catch (BadLocationException e) {
+            e.printStackTrace();
+        }
+
+        // getAllEnabledBy would return all transitive features but typically offers too much
+        Set<String> featureCandidates = LibertyProjectsManager.getInstance()
```

Comment:
Is it possible for this Set to be empty? Just wondering if we should make sure it is non-empty before we spend all the effort of figuring out the text to enter for the quick fix above?